pkg/client/patch.go

```diff
@@ -211,3 +212,55 @@ func (p applyPatch) Data(obj Object) ([]byte, error) {
 	// client-go does, more-or-less).
 	return json.Marshal(obj)
 }
+
+// applyFromPatch uses server-side-apply with a ApplyConfiguration object.
+type applyFromPatch struct {
+	patch interface{}
+}
+
+// Type implements Patch.
+func (p applyFromPatch) Type() types.PatchType {
+	return types.ApplyPatchType
+}
+
+// setNameNamespace sets name and namespace for the returned object if an ApplyConfiguration was supplied.
+func setNameNamespace(obj Object, ac interface{}) error {
+	if !reflect.ValueOf(ac).Elem().IsValid() || reflect.ValueOf(ac).Elem().Kind() != reflect.Struct {
+		return fmt.Errorf("ApplyConfiguration is not a valid struct")
+	}
+	// Since Name and Namespace are embedded objects, nil checks must be
+	// performed against the embedded ObjectMeta object first.
+	if reflect.ValueOf(ac).Elem().FieldByName("ObjectMetaApplyConfiguration").IsNil() {
```

Comment:
Instead of using an empty interface and reflect, can we define an interface with getters/setters fro the properties we need here (if it doesn't exist already)?

That would also make the whole thing a lot more intuitive to use and give at least some compile-time verification

Reply:
I thought about doing this and it's easy to accomplish with just the CRD types. However, the apply configurations for native k/k resources don't implement the interface at the moment. Since it takes some time for k/k changes to propagate, would it be possible to keep this as is and introduce the getters/setters in a future version?

Reply:
As far as I can tell, the existing `Patch` methods derive the namespace from the supplied `Object`, not from the patch content—at least for ["typed" objects](https://sourcegraph.com/github.com/kubernetes-sigs/controller-runtime@v0.11.1/-/blob/pkg/client/typed_client.go?L125) and [unstructured objects](https://sourcegraph.com/github.com/kubernetes-sigs/controller-runtime@v0.11.1/-/blob/pkg/client/unstructured_client.go?L161), anyway.

Reply:
As far as I can tell, the existing `Patch` methods derive the namespace from the supplied `Object`, not from the patch content—at least for ["typed" objects](https://sourcegraph.com/github.com/kubernetes-sigs/controller-runtime@v0.11.1/-/blob/pkg/client/typed_client.go?L125) and [unstructured objects](https://sourcegraph.com/github.com/kubernetes-sigs/controller-runtime@v0.11.1/-/blob/pkg/client/unstructured_client.go?L161), anyway.